:book: describe need of RBAC permissions to allow Prometheus Operator to have access to the metrics (#2827)
kubernetes-sigs/kubebuilder: docs/book/src/reference/metrics.md

monitor-controller-manager-metrics-monitor   2m8s
```

> 

Reviewer: ```suggestion
<aside class="note">
<h2>RBAC for Prometheus</h1>
```

Author: added

Note that you should ensure that RBAC roles scaffold in the `config/prometheus`
were applied to the namespace where the manager is deployed. Otherwise,
Prometheus server will not be able to access the ServiceMonitor that describes
the set of targets to be monitored by Prometheus.

When using [prometheus-operator](https://github.com/prometheus-operator) for
cluster monitoring, more information about the RBAC requirements can be found in
the [official documentation](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/rbac.md).

> 

Reviewer: ```suggestion
</aside>

```

Author: added

monitor-controller-manager-metrics-monitor   2m8s
```

### RBAC for Prometheus

If you are using
[kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) or
[prometheus-operator](https://github.com/prometheus-operator/prometheus-operator)
for cluster monitoring and you have trouble seeing your manager metrics in
Prometheus, check the Prometheus Targets page for the resource 
`serviceMonitor/<namespace>/<project-prefix>-controller-manager-metrics-monitor`.

If you can't find it, you need to give extra permissions to the Prometheus
ServiceAccount.

> **⚠ WARNING**  

> > A ClusterRole for the Prometheus `prometheus-k8s` ServiceAccount in the

Reviewer 2: By default, A ClusterRole will be created for the Prometheus `prometheus-k8s` ServiceAccount in the
`monitoring` namespace. Please ensure that the ServiceAccount exists and that it belongs to Prometheus. If the name or namespace ?of Prometheus? are different, you should change them in
`config/prometheus_rbac/cluster_role_binding.yaml` before applying.

"Please ensure that the ServiceAccount exists and that it belongs to Prometheus" - what does it mean? How do they make sure of that?

monitor-controller-manager-metrics-monitor   2m8s
```

### RBAC for Prometheus

When using [prometheus-operator](https://github.com/prometheus-operator) for
cluster monitoring, you should ensure that necessary RBAC roles and permissions
were applied to the namespace where the manager is deployed. Otherwise,
Prometheus server will not be able to access the ServiceMonitor that describes
the set of targets to be monitored by Prometheus.

> ```yaml

Reviewer: Do we need to add it here? is not that in the Prometheus Operator doc? 
Is not better only so that if they cange that we do not get this doc outdated?

> apiVersion: rbac.authorization.k8s.io/v1

Reviewer: Do we need to add it here? is not that in the Prometheus Operator doc? 
Is not better only so that if they cange that we do not get this doc outdated? 

What about a note with "warning" (see that will render very nice on the docs)

```
<aside class="warning">
<h2>RBC required for Prometheus Operator have access to the metrics</h2>

If you have not Promethues installed on the cluster and you are using Promethues 
Operator then be aware that when using [prometheus-operator][rbac-promethues-operator] for
cluster monitoring, you should ensure the necessary RBAC roles and permissions. Otherwise,
Prometheus server will not be able to access the ServiceMonitor that describes
the set of targets to be monitored by Prometheus.

</aside>

Also, note that these are ClusterRoles so does not need to be applied on the ns of the operator.

```

Reviewer 2: Hi @camilamacedo86 @joelanford, I think that we should provide this as an optional tool and provide an api call for it.
In the documentation we will explain when it can be used and also enable changing the namespace and additional details that may change. The namespace and service account name would have a default, but we should add to the tool flags to update them.

Reviewer: @sradco is specific for Prometheus Operator and Prometheus Operator describes how to do that, see: https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/rbac.md

Therefore I do not think that Kubebuilder should be responsible for providing and maintaining helpers over "how to configure Prometheus Operator itself". The RBAC proposed to allow Prometheus Operator to have access to the ServiceMonitor used by each operator. However, users might have it set up already or prefer as @joelanford described allow this permission to specific SA instead.  

We here could/should highlight the "pre-requirement" to check the metrics with Prometheus Operator since we are suggesting its usage, see the metrics docs:

> Install Prometheus and Prometheus Operator. We recommend using [kube-prometheus](https://github.com/coreos/kube-prometheus#installing) in production **if you don’t have your own monitoring system**

Also, note that we are saying in the doc as well: 

> To achieve that, you can create a clusterRoleBinding to bind the clusterRole to the service account that your Prometheus server uses. 

Therefore, we just need a warning note pointing out the users to ensure that **if** they have Prometheus installed via the Operator then they need to ensure that Prometheus Operator has the required permissions RBAC to access the ServiceMonitor in the namespace where the Operator will be deployed and should link their documentation as well. ( We should not doc how Prometheus Operator works or how to config it, it is also part of the Prometheus Operator domain. )

 

Because

monitor-controller-manager-metrics-monitor   2m8s
```

<aside class="warning">
<h2>If you are using Prometheus Operator ensure that you have the required
permissions</h2>

If you are using Prometheus Operator then be aware that by default, its RBAC
rules are only enabled for the `default` and `kube-system namespaces`. See its
guide to know how to configure the
[kube-prometheus](https://github.com/prometheus-operator/kube-prometheus/blob/main/docs/monitoring-other-namespaces.md)
to monitor other namespaces with `.jsonnet`.

> However, you can also enable it using RBAC. See the Prometheus Operator

Reviewer 3: ```suggestion
Alternatively, you can give the Prometheus Operator permissions to monitor other namespaces using RBAC. See the Prometheus Operator
```

> [Getting Started](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/getting-started.md#enable-rbac-rules-for-prometheus-pods)

Reviewer 2: @machadovilaca for some reason when I press the link it doesn't get me to the |"Enable RBAC rules for Prometheus pods" section, eventough your link it to it. Can we change the name of the link from "Getting Stared" to "Enable RBAC rules for Prometheus pods" ?

Reviewer: +1

Author: updated

documentation to know how to enable the permissions on the namespace where the
`ServiceMonitor` and manager are applied.
</aside>

> 

Reviewer: @alculquicondor WDYT?

> `ServiceMonitor` and manager are applied.

Reviewer 3: ```suggestion
`ServiceMonitor` and manager exist.
```

monitor-controller-manager-metrics-monitor   2m8s
```

3. If you are using the suggested
[kube-prometheus](https://github.com/coreos/kube-prometheus#installing) as a
monitoring system, and you want Prometheus Operator to have access to the
controller metrics, create the necessary RBAC resources by applying the
following `Role` and `RoleBinding` for Prometheus to be able to monitor our

> controller namespace.

Reviewer: ```suggestion
<aside class="warning">
<h2>If you are using Prometheus Operator ensure that you have the required permissions </h2>

If you are using [Prometheus Operator][rbac-promethues-operator] then be aware that by default, its RBAC rules are only enabled for the Default and kube-system namespaces. See
its documentation about how to configure the [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus/blob/main/docs/monitoring-other-namespaces.md) to monitor another
namespaces with `.jsonnet`.

However, you can also enable it using RBAC. See the Prometheus Operator [Getting Started](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/getting-started.md#enable-rbac-rules-for-prometheus-pods) documentation to know how to enable the permissions on the namespace where the ServiceMonitor and manager are applied.
</aside>

Reviewer: @machadovilaca,

(imo) we need to link here the Prometheus Operator documentation.
See that it will help the users know more about if they need it as well.

On top of that, `<aside class="warning">`  will create a note/CSS on the docs that can retain the attention of the users.

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1

> kind: Role
> metadata:
>    name: <project-prefix>-controller-manager-cluster-role
>    namespace: <namespace>
> rules:
>    - apiGroups: [""]
>      resources:
>         - services
>         - endpoints
>         - pods
>      verbs: ["get", "list"]
> ---
> apiVersion: rbac.authorization.k8s.io/v1
> kind: RoleBinding
> metadata:
>    name: <project-prefix>-controller-manager-cluster-role-binding

Reviewer: ```suggestion
```

>    namespace: <namespace>
> roleRef:
>    apiGroup: rbac.authorization.k8s.io
>    kind: Role
>    name: <project-prefix>-controller-manager-cluster-role
> subjects:
>    - kind: ServiceAccount
>      name: <prometheus-service-account>
>      namespace: <prometheus-service-account-namespace>
> ```
> 

Reviewer: ```suggestion
```
